Allow reauthentication if the OIDC state cookie is not matched

sberyozkin opened the pull request:
Fixes #34630

sberyozkin committed:
Allow reauthentication if the OIDC state cookie is not matched

michalvavrik commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java:
```diff
@@ -859,6 +859,33 @@ public enum ResponseMode {
+         * It will cause a new authentication redirect to OpenId Connect provider.
+         * <p/>
+         * Make sure it won't cause browser redirect loops if you decide to do so.
+         * Please keep in mind that unmatched state cookies will have to be removed.
```
By this, you probably mean that Quarkus will remove all state cookies when `failOnMissingStateParam` is `false` and there is zero or many state query params, right? Before I read the code, I thought that client need to keep in mind that response will still contain state cookies. Is sentence clear to you @sberyozkin , please?

michalvavrik replied:
Hmm, maybe I wasn't very clear. What I mean is that I would add removed _by whom_.

sberyozkin replied:
Indeed, it should be `Please keep in mind that unmatched state cookies will have to be removed if this property is set to false`

>and there is zero or many state query params

At the moment this PR is only about supporting a case where a request arrives, we have a state cookie, but no `state` query parameter. At the moment it is an immediate `401` but I'd like to give users an option to ignore that the state cookie has not been matched and just request reauthentication - and because the state cookie is not matched, it has to be cleared.

The reason I add the text cautioning about it is, if we have a multi-tab authentication, example, an authentication is in process in one tab, user is challenged to authenticate by Keycloak, does not complete it, then opens another tab, a new request is issued, here we have another state cookie. But this multi-tab authentication is not even supported right now, without this PR, it works only if the request URL and the callback URLs are different.

So I'll probably just remove this caution to avoid the confusion, I've nearly got confused myself while typing the above text :-)

michalvavrik replied:
Alright, I understand now. Thank you

michalvavrik commented on extensions/oidc/runtime/src/main/java/io/quarkus/oidc/OidcTenantConfig.java:
```diff
@@ -859,6 +859,33 @@ public enum ResponseMode {
         public boolean allowMultipleCodeFlows = true;
 
+        /**
+         * Fail with the HTTP 401 error if the state cookie is present but no state query parameter is present.
```
> no state query parameter is present

It also impacts state when more than one state query param is present. You don't mention it because it doesn't matter - if there is multivalued state query param, you can't match it with the cookie, correct?

sberyozkin replied:
Only a single value state query parameter is supported, I'll need to add a dedicated handling and failure for a multiple-valued state query

michalvavrik replied:
Ok, thx

michalvavrik reviewed (commented):
The PR LGTM, but I have 2 questions to docs to ensure it is clear to users.

sberyozkin commented:
@michalvavrik Thanks, your reviews are welcome, you have done a real lot of security work for Quarkus

sberyozkin commented:
@michalvavrik I've removed the text from JavaDocs which may be confusing, and also added a test confirming that something like `state=1&state=2` with the available state parameter causes 401 - a case like this it can't happen by accident so deserves 401. Minor update to the code to check the new property if the state cookie is not matched with the available single state parameter - given that state cookies are unique it might happen, so instead of 401 users should be able to request re-authentication instead. If you have any comments - let me know.

@gastaldi @pedroigor Can you please review as well, as Michal's approval does not enable a Merge button yet
